doc/manual/contract.md

```diff
@@ -0,0 +1,505 @@
+
+Methods, as well as attribute setters, can be associated with two types of contracts; precondition (`expect`) that expresses requirements of the routine´s call and postcondition (`ensure`) that expresses what the routine guarantee after her execution. Together, they define the rules to be respected between the callee and the caller. If the caller fulfills the `expect` condition on entry then the caller can assume that the callee fulfill the `ensure` condition on exit.
+
+All contracts (`expect`, `ensure`) are evaluated in the same context as the called method, it's possible to refer to attributes, methods and parameters in the condition.
+
+When defining a contract (`expect`, `ensure`) on an attribute like bellow, only the automatically generated write property (setter) will use the contract.
+
+~~~
+var baz: Int is expect(baz > 0) # In the contract, baz refers to the parameter of the write property (setter).
+~~~
+
+### Define preconditions with `expect`
+
+We define a precondition using the annotation `expect`. In the `Stack` class, the method `pop` defines a precondition, that a call to the method should be made on a non-empty stack.
+
+~~~
+fun pop is expect(not_empty)
+~~~
+
+### Define postconditions with `ensure`
+
+We define a postcondition using the annotation `ensure`. In the `Stack` class, the method `push` defines postconditions. The first guarantee defines that, the method must add the element received as argument (`object`) at the top of the stack.
+
+~~~
+fun push(object: T) is ensure(self[size - 1] == object)
+~~~
+
+Within an `ensure` contract we can refer to the return value of the method with the keyword `result`. In the `Stack` class, the `[]` method guarantees that, the return value is the same as the element stored in the stack at the given index.
+
+~~~
+fun [](index: Int): T is ensure(result == data[index])
+~~~
+
+Note that the keyword `result` can only be used when the method has a return parameter.
+
+
+In `ensure` we can refer to any arguments and instance variables before the method call with the using of `old` keyword. Again, with the `pop` method, we can ensure that the `size` of the stack after the method call will be the same as before minus one `old(size) - 1 == size`.
+
+~~~
+fun pop(object: T) is ensure(old(size) - 1 == size)
+~~~
+
+In an `old(expression)` the `expression` is evaluated in the same context as the called method. Thus, it's possible to refer to attributes, methods, and arguments.
+
+`old(object)` stores a reference to object. By doing so, if `object` is mutated in the method the `old(object)` will also be mutated.
+
+Let us take a look on the following example:
+
+~~~
+class Boat
+	var position: Position
+
+	fun move(direction: Position)
+	do
+		position =  position + direction
+	end
+end
+
+class Position
+	var x: Float
+	var y: Float
+
+
+	fun +(position: Position): Position
+	is
+		ensure(old(x) + position.x == x)
+		nit ensure(old(y) +  position.y == y)
+	do
+		_x += position.x
```

Comment:
What's up with the `_` it doesn't look like the way we traditionally write Nit code?

Reply:
`_` is used to access the attribute without using the write or read property. On the other hand, it's not a good practice to put in an example.